language/stdlib/modules/libra_system_2.move

```diff
@@ -0,0 +1,264 @@
+        // Ensure that this address is not already a validator
+        Transaction::assert(!is_validator_(account_address, &validator_set_ref.validators), 18);
+
+        let config = ValidatorConfig2::get_config(*account_address);
+        let current_block_time = LibraTimestamp::now_microseconds();
+        Vector::push_back(&mut validator_set_ref.validators, ValidatorInfo {
+            addr: *account_address,
+            config: *&config, // copy the config over to ValidatorSet
+            consensus_voting_power: 1,
+            last_force_update_time: current_block_time,
+        });
+        
+        emit_reconfiguration_event();
+    }
+
+    // Removes a validator, only callable by the LibraAssociation address
+    public fun remove_validator(account_address: &address) acquires ValidatorSet {
+        Transaction::assert(Transaction::sender() == 0xA550C18, 1);
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        let size = Vector::length(&validator_set_ref.validators);
+        // Make sure the size of validator set does not go beyond 4, which is the minimum safe number of validators
+        Transaction::assert(size >= 5, 22);        
+        
+        // Ensure that this address is an active validator
+        Transaction::assert(is_validator_(account_address, &validator_set_ref.validators), 21);
+
+        // Remove corresponding ValidatorInfo from the validator set
+        let to_remove_index = get_validator_index_(&validator_set_ref.validators, account_address);
+        _  = Vector::swap_remove(
+            &mut validator_set_ref.validators,
+            to_remove_index
+        );
+        emit_reconfiguration_event();
+    }
+    
+    // Copies validator's config to ValidatorSet/
+    // Callable by the LibraAssociation address only.
+    // Returns true if some portion of the config was changed and false otherwise
+    // Aborts if there is no change to the config.
+    public fun update_validator_info(addr: &address) acquires ValidatorSet {
+        Transaction::assert(Transaction::sender() == 0xA550C18, 1);
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+
+        Transaction::assert(is_validator_(addr, &validator_set_ref.validators), 19);
+        let i = get_validator_index_(&validator_set_ref.validators, addr);
+
+        Transaction::assert(update_ith_validator_info_(&mut validator_set_ref.validators, i), 42);
+        emit_reconfiguration_event();
+    }
+
+    // This function can be called by anyone in posession of a valid ValidatorConfigUpdateToken,
+    // only validators or validators' delegated addresses can get this token as an output of
+    // rotate_consensus_key function.
+    // Returns true if the config was updated and false if config stays unchanged
+    public fun force_update_validator_info(update_token: ValidatorConfig2::ValidatorConfigUpdateToken) acquires ValidatorSet {
```

Comment:
What would go wrong if you took `validator_address` as input here instead of an update token?

As I understand it, the purpose of the update token is to allow either the validator or the delegate to force a key rotation. But couldn't you either:
- Directly assert that the sender is the validator or delegate? (which I think accomplishes the same thing without requiring the indirection of `ValidatorConfigUpdateToken`)?
- [if this is permissible] Allow any validator (or any delegate) to force an update? This function already aborts at line 134 if there hasn't been a rotation for `validator_address`, so the potential harm here seems minimal

Reply:
These are such good questions!
1) Yeh, we can assert that the sender is the validator or delegate - I was trying to make LibraSystem oblivious to the fact that we have delegation as I was thinking that in the future we could decide on some even more complicated validator's key-management logic, so with this approach we won't need to change the LibraSystem module. But getting rid of the update_token will simplify the code for sure. So, do you think asserting that the sender is the validator or delegate instead of having this update_token, is a good enough way to do it for now? Either is fine with me...

2) Allow any validator to update should not cause troubles, but it's a little weird: say the regular update happens once every month, so I, as a validator, prepare for it and rotate my key a week in advance. I still expect my new key to become active in a week, but suddenly someone else makes it active (even though I didn't ask for a force-update). It would be nice to give each validator control over which key to use...

Reply:
So I changed to the simpler API that you suggested, @sblackshear , and removed the update token for now, we can revisit it later!